# LTA detection

This repository contains scripts for the detection and analysis of Loss-Translocation-Amplification chromothripsis events to complement the manuscript "Mechanisms underpinning osteosarcoma genome complexity and evolution" by Valle-Inclan et al. 

*   Valle-Inclan JE, De Noon S, Trevers K, Elrick H, Tanguy M, Butters T, et al. Mechanisms underpinning osteosarcoma genome complexity and evolution. bioRxiv. 2023. p. 2023.12.29.573403. doi:10.1101/2023.12.29.573403 


## Usage

Required input generated by the HMF pipeline (https://github.com/hartwigmedical/hmftools)

* SVs from GRIDSS
* Copy number segments from PURPLE
* Driver alterations 

Additional input:

* Cohort file
* Chromosome arm coordinates

Please specify paths in the cohort file `lta_detection.conf`

Optional arguments: `show_plots=FALSE` to skip recon plots; `dataset_selection_label='TCGA-SARC'` to subset the cohort file and `lta_tsg_disruption_lst=c("_TP53_")` to limit analysis to specified TSGs. 

`SINGULARITYENV_R_MAX_VSIZE=150Gb
singularity exec --bind /path/to/input/ /path/to/script/structural_variation_202405_amd.sif R -e "source('/path/to/script/lta_detection.conf ');show_plots=T;dataset_selection_label='TCGA-SARC';source('/path/to/script/lta_detection.R')"`

Plots and tables

`singularity exec --bind /path/to/input/ /path/to/script/structural_variation_202405_amd.sif R -e "source('/path/to/script/lta_detection.conf ');source('/path/to/script/lta_analysis.R')"`


Image: https://hub.docker.com/layers/ivanbelzen/structural_variation/202405_amd/images/sha256-19b705ec12273cb4b01dcb46da378e5fb54308692772e000414c09d44a6fa23d

Also see example run script: `run_lta_detection.sh`

## Approach for detecting LTA events 

The main characteristic of the LTA events discovered in osteosarcoma is a double strand break leading to concomitant _TP53_ inactivation and segmental amplifications, often harbouring oncogenes. This process is mediated by unstable dicentric chromosomes and contributes to genome-wide genomic instability and the formation of highly complex karyotypes. To detect LTA events, we formulated criteria to look for evidence of biallelic _TP53_ disruption co-occurring with dicentric chromosome formation. 

* Biallelic _TP53_ disruption: driver mutations identified by the HMF pipeline in addition to SV breakpoints inside _TP53_ and loss of heterozygosity (LOH).
* Connections between chromosome 17p and other chromosome arms with CGRs.
* Presence of an SV in the _TP53_ gene body, or a translocation breakpoint or a CGR in the region between start of _TP53_ and the centromere.
* Loss of the 17p terminal segment, defined as the genomic region from the start of the chromosome up to and including _TP53_: 
   * minimum of 75% LOH
   * Alternatively: minimum 33% LOH together with an amplification (minimum 1kbp of 9 copies or more)
* Absence of full chromosome LOH and chromosome arm LOH
   * Removing cases where the chromosome (arm) shows extensive LOH (>90%) and exists in a single copy number state (>90%). This removes aneuploidies but does not discard extensive LOH due to rearrangements.
   * Removing all cases with full chromosome arm LOH (>99.5%)

This filtering removes cases where the gene is knocked-out by a simple SV in a completely haploid chromosome arm, as well as non-specific losses of the entire 17p chromosome arm. However, it allows for cases where the terminal segment loss is obfuscated due to focal amplifications or non-specific LOH of the chromosome (arm) is present due to additional rearrangements.

To systematically detect LTA events across cancer types, we expanded the analysis to include disruption of other TSGs. Since we observed in osteosarcoma that LTA events with _TP53_ can also affect other TSGs, rather than those events being independent, we annotated these non-TP53 LTAs were annotated with whether they are connected to the _TP53_ gene region. Furthermore, we annotated LTA events with partner chromosomes that harbour oncogene amplifications to enable selection of events that potentially provide a selective advantage to the tumor. 



# References

* LTA in osteosaroma
   *   Valle-Inclan JE, De Noon S, Trevers K, Elrick H, Tanguy M, Butters T, et al. Mechanisms underpinning osteosarcoma genome complexity and evolution. bioRxiv. 2023. p. 2023.12.29.573403. doi:10.1101/2023.12.29.573403 

* ReConPlot
   *   Valle-Inclan, JE, Cortes-Ciriano I, ReConPlot: an R package for the visualization and interpretation of genomic rearrangements, Bioinformatics. 2023, doi: 10.1093/bioinformatics/btad719

* HMF Pipeline
   *   https://github.com/hartwigmedical/hmftools

* GRIDSS
   *   Cameron DL, Schröder J, Penington JS, Do H, Molania R, Dobrovic A, Speed TP, Papenfuss AT (2017) GRIDSS: sensitive and specific genomic rearrangement detection using positional de Bruijn graph assembly. Genome Research 27:2050–2060

* PURPLE
   * Priestley P, Baber J, Lolkema MP, Steeghs N, de Bruijn E, Shale C, et al. Pan-cancer whole-genome analyses of metastatic solid tumours. Nature. 2019;575: 210.

*  Chromosome bands retrieved from the UCSC table browser: Navarro Gonzalez J, Zweig AS, Speir ML, et al (2020) The UCSC Genome Browser database: 2021 update. Nucleic Acids Res 49:D1046–D1057

